### PR TITLE
[feat] Expose Keccack Raw Inputs in Bytes instead of Input RLCs

### PR DESCRIPTION
**Pull request by nyunyunyunyu**
## Description
Add column `word_value` `bytes_left` to represent raw inputs in 8 bytes. Most are adopted from [this PR](https://github.com/scroll-tech/zkevm-circuits/pull/216).
- For every non-padding round, `word_value` and `bytes_left` on its 1st row represents the input word of this round(8 bytes) and the number of input bytes left.
```
data:
                get_num_rows_per_round: 18
                input: "12345678abc"
            table:
                Note[1]: be careful: is_paddings is not column here! It is [Cell; 8] and it will be constrained later.
                Note[2]: only first row of each round has constraints on bytes_left. This example just shows how witnesses are filled.
        offset word_value bytes_left  is_paddings q_enable q_padding_last
        18     0x87654321    11          0         1        0 // 1st round begin
        19        0          10          0         0        0
        20        0          9           0         0        0
        21        0          8           0         0        0
        22        0          7           0         0        0
        23        0          6           0         0        0
        24        0          5           0         0        0
        25        0          4           0         0        0
        26        0          4           NA        0        0
        ...
        35        0          4           NA        0        0  // 1st round end
        36      0xcba        3           0         1        1  // 2nd round begin
        37        0          2           0         0        0
        38        0          1           0         0        0
        39        0          0           1         0        0
        40        0          0           1         0        0
        41        0          0           1         0        0
        42        0          0           1         0        0
        43        0          0           1         0        0
```
## Side Effects
- Remove `challenge`. Now Keccak circuits don't have phase 2.
- Remove `input_rlc` and `input_len` columns.
- Change `MAX_DEGREE` from 3 to 4 in order to support new gates.
- Rename `q_padding` to `q_input`. `q_padding` is confusing because it actually means a round has inputs with paddings.

## Testing Plan
Unit testing

### CODE REVIEW

**Comment by jonathanpwang** [comment]
Please reference the Scroll PR this is related to.

And just checking, the plan is still to after this go from bytes to something like `u64` or `u128` right?

**Comment by nyunyunyunyu** [comment]
> Please reference the Scroll PR this is related to.
> 
> And just checking, the plan is still to after this go from bytes to something like `u64` or `u128` right?

Yes I will do it in halo2-lib.

**Comment by jonathanpwang** [comment]
Update PR description with new input format please

**Comment by nyunyunyunyu** [comment]
> Update PR description with new input format please

Added

**Comment by jonathanpwang** [comment]
Note to myself: I realized the naming of `q_padding` and `is_padding` is confusing and refer to different things. `q_padding` is fixed and are the rounds where there is no absorb, while `is_padding` refers to the actual padding of the input bytes...

File: hashes/zkevm/src/keccak/mod.rs [hunk 1]
```diff
@@ -568,6 +553,111 @@ impl<F: Field> KeccakCircuitConfig<F> {
...
+            cb.condition(q_in_round(q_first, meta), |cb| {
+                cb.require_zero(
+                    "bytes_left needs to be zero on the first row",
+                    meta.query_advice(keccak_table.bytes_left, Rotation::cur()),
+                );
+            });
+            // is_paddings only be true when
+            cb.condition(q(q_padding, meta), |cb| {
+                for i in 0..num_rows_per_round {
+                    let bytes_left = meta.query_advice(keccak_table.bytes_left, Rotation(i as i32));
```
Comment (jonathanpwang) [inline hashes/zkevm/src/keccak/mod.rs:616]: As discussed in `tests.rs`, this is still recording `bytes_left` for each input byte, which means `rows_per_round >= 8` assumption is necessary. 

I think it is possible, without increasing custom gate `MAX_DEGREE` to just update `bytes_left` once per round by summing over all of `is_paddings` in this custom gate?

File: hashes/zkevm/src/keccak/mod.rs [hunk 2]
```diff
@@ -568,6 +553,111 @@ impl<F: Field> KeccakCircuitConfig<F> {
...
+                        "bytes_left should be 0 when padding",
+                        bytes_left * is_paddings[std::cmp::min(i, NUM_BYTES_PER_WORD - 1)].expr(),
+                    );
+                }
+            });
+            cb.condition(
+                (q_in_round(q_enable, meta)
+                    - q_in_round(q_padding, meta)
+                    - q_in_round(q_first, meta))
+                    * meta.query_advice(is_final, Rotation::next()),
```
Comment (jonathanpwang) [inline hashes/zkevm/src/keccak/mod.rs:642]: Is this the right condition? Just checking.

This is saying, in an absorb round which is not a dummy round, on the row before the `is_final`, enforce equality. `is_final` is enforced to only be the first row of a round, and it _should_ only be in the last round.

**Review by jonathanpwang — CHANGES_REQUESTED** [review]
Main concerns: `bytes_left` requiring `rows_per_round >= 8` and the `word_value` gate constraint.

It would also be helpful to have a more extensive example detailing what happens at padding and is_final rounds.

File: hashes/zkevm/src/keccak/mod.rs [hunk 3]
```diff
@@ -568,6 +553,111 @@ impl<F: Field> KeccakCircuitConfig<F> {
...
+                for i in 0..num_rows_per_round {
+                    let bytes_left = meta.query_advice(keccak_table.bytes_left, Rotation(i as i32));
+                    let bytes_left_next =
+                        meta.query_advice(keccak_table.bytes_left, Rotation(i as i32 + 1));
+                    // real input only when !is_paddings[i] && i < NUM_BYTES_PER_WORD
+                    cb.require_equal(
+                        "if not padding, bytes_left decreases by 1, else, stay same",
+                        bytes_left_next,
+                        bytes_left.clone()
+                            - if i < 7 || i == num_rows_per_round - 1 {
```
[the remark this answers is not in the document]
Comment (nyunyunyunyu) [reply hashes/zkevm/src/keccak/mod.rs:624]: This part has been rewritten.

File: hashes/zkevm/src/keccak/mod.rs (hunk 2 above)
Comment (nyunyunyunyu) [reply hashes/zkevm/src/keccak/mod.rs:643]: This part has been rewritten.

File: hashes/zkevm/src/keccak/mod.rs (hunk 1 above)
Comment (nyunyunyunyu) [reply hashes/zkevm/src/keccak/mod.rs:616]: This part has been rewritten.

File: hashes/zkevm/src/keccak/mod.rs [hunk 4]
```diff
@@ -567,6 +552,111 @@ impl<F: Field> KeccakCircuitConfig<F> {
...
+        41        0          0           1         0        0
+        42        0          0           1         0        0
+        43        0          0           1         0        0
+        */
+
+        meta.create_gate("word_value", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+            cb.require_equal(
+                "word value",
+                absorb_data.expr(),
```
[the remark this answers is not in the document]
Comment (nyunyunyunyu) [reply hashes/zkevm/src/keccak/mod.rs:599]: This part has been rewritten.

File: hashes/zkevm/src/keccak/mod.rs [hunk 5]
```diff
@@ -568,6 +555,103 @@ impl<F: Field> KeccakCircuitConfig<F> {
...
+                meta.query_advice(keccak_table.word_value, Rotation::cur()),
+            );
+            cb.gate(q(q_input, meta))
+        });
+        meta.create_gate("bytes_left", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+            let bytes_left_expr = meta.query_advice(keccak_table.bytes_left, Rotation::cur());
+
+            // bytes_left is 0 in the absolute first `rows_per_round` of the entire circuit, i.e., the first dummy round.
+            cb.condition(q_in_round(q_first, meta), |cb| {
```
Comment (jonathanpwang) [inline hashes/zkevm/src/keccak/mod.rs:622]: Can this just be `q(q_first)`? The entire `cb` has `cb.gate(q_enable)` so I think it's only on the first row anyways. If that's true, please update commend above slightly.

**Review by jonathanpwang — APPROVED** [review]
LGTM!
Just left one final comment making sure `bytes_left` is only ever constrained on first row of each round. Can merge whenever you feel good.

**Pull request merged**